### PR TITLE
Embed fnc spec as annotation to detect diff

### LINKED ISSUE #47: Deployments keep being updated when docker registry secret is provided
I am on openfaas-operator 0.8.6 and I am deploying a function which its image is in a private registry.

I followed https://docs.openfaas.com/deployment/kubernetes/#create-an-image-pull-secret to provide the registry secret in my `stack.yml`. The `faas-cli deploy` had been accepted, and the deployment was created successfully in Kubernetes.

But I found that there were lots of exited Pods of this function, and the Pods got terminated and recreated again and again, very frequently. I checked the log of the operator, here is what I got

```
Function my-func secrets have changed
Updating deployment for my-func
Function my-func secrets have changed
Updating deployment for my-func
Function my-func secrets have changed
Updating deployment for my-func
Function my-func secrets have changed
Updating deployment for my-func
...
```

The operator was looping that 2 lines (`Function...; Updating...`) all the time, with a short interval. Obviously the deployment kept being updated, leading to the pods' terminations and recreations.

The first line of log comes from https://github.com/openfaas-incubator/openfaas-operator/blob/0.8.6/pkg/controller/deployment.go#L234

And the second line comes from https://github.com/openfaas-incubator/openfaas-operator/blob/0.8.6/pkg/controller/controller.go#L316 

Behind the first line, there is a checking if the secrets between the function and the deployment are equal, by comparing the slice `function.Spec.Secrets` and the slice built by `deployment.Spec.Template.Spec.Volumes[n].Projected.Sources.Secret.Name`

In this case,

* This checking failed
* The log `Function my-func secrets have changed` (first line) was printed
* `deploymentNeedsUpdate` returned `true`
* The log `Updating deployment for my-func` (second line) was printed
* Deployment was updated

So I did the below to see why the checking failed.

First I inspected the function
```bash
$ kubectl -n func get functions my-func -o yaml
```
I found
```
  secrets:
  - my-registry-secret
```

The I inspected the deployment
```bash
kubectl -n func get deployment my-func -o yaml
```
I found
```
      imagePullSecrets:
      - name: my-registry-secret
```
But I didn't find any `volumes`, and this is the problem.

According to https://github.com/openfaas-incubator/openfaas-operator/blob/0.8.6/pkg/controller/secrets.go#L31, if the secret type is `corev1.SecretTypeDockercfg` or `corev1.SecretTypeDockerConfigJson`, the secret will be appended to `deployment.Spec.Template.Spec.ImagePullSecrets`, and nothing for `deployment.Spec.Template.Spec.Volumes`

Here my `my-registry-secret` got appended to `imagePullSecrets` in the deployment, causing the absence of `volumes`, this produced non-empty `function.Spec.Secrets` (with 1 item `my-registry-secret`), and an empty `deployment.Spec.Template.Spec.Volumes`, leading to the checking failure.

I think this problem will exist as long as you have registry secret provided in the `stack.yml`, no matter you have application secret or not. The registry secret will be always in `function.Spec.Secrets`, and always not in `deployment.Spec.Template.Spec.Volumes`, the checking will always fail.

I have added the following snippet before https://github.com/openfaas-incubator/openfaas-operator/blob/0.8.6/pkg/controller/deployment.go#L233 to remove all the ImagePullSecrets, and so far it is working.

```go
	var secrets []string
	secrets = append(secrets, function.Spec.Secrets...)
	if len(secrets) > 0 && len(deployment.Spec.Template.Spec.ImagePullSecrets) > 0 {
		for i := len(secrets) - 1; i >= 0; i-- {
			for _, s := range deployment.Spec.Template.Spec.ImagePullSecrets {
				if s.Name == secrets[i] {
					secrets = append(secrets[:i], secrets[i+1:]...)
					break
				}
			}
		}
	}

	if secretsNotEqual(secrets, deployment.Spec.Template.Spec.Volumes) {
		glog.V(2).Infof("Function %s secrets have changed",
			function.Spec.Name)
		needsUpdate = true
	}
```

### PR DESCRIPTION
**Pull request by LucasRoesler**
**What**
- Add annotation to embed the function spec from the CRD into the
Deployment object that we create
- Use the embedded function spec to determine if the function has
changed since we last modified it
- Update each check to only to reference the embedded spec

**Why**
- This avoids potential conflict with other controllers that might edit
the object by only watching that diff in the spec

Fixes #47

### CODE REVIEW

**Comment by LucasRoesler** [comment]
Sorry for the delayed response, I just pushed `theaxer/openfaas-operator:fix47` based on this branch

**Comment by stefanprodan** [comment]
I've tested the upgrade and all deployments are recreated with the annotation in place. @LucasRoesler have you tested the image pull secret use case?

**Comment by LucasRoesler** [comment]
I tested 4 different secrets cases. It would still be nice to get validation from one more person though

**Comment by alexellis** [comment]
Hi Lucas, thank you for working on this 💯 🥇 

Quick question for you - is the `annotationFunctionSpec` going to be returned on the /system/functions /system/function/<fn> routes via the API?

**Comment by alexellis** [comment]
I think it should be hidden/removed from the user.

**Comment by alexellis** [comment]
I deployed your image to my local cluster and saw something odd, hopefully @stefanprodan can comment?

```
[{"name":"certinfo","image":"stefanprodan/certinfo","invocationCount":0,"replicas":1,"envProcess":"","availableReplicas":1,"labels":null,"annotations":null}]
```

Labels should never be null.

Here are the labels I see on Swarm (by default):

```
"labels":{"com.openfaas.function":"ruby-echo","com.openfaas.uid":"603975054","function":"true"}
```

**Comment by alexellis** [comment]
We are also missing annotations.

This is what Kubernetes gives with faas-netes:

```
"labels":{"faas_function":"figlet2"},
"annotations":{"prometheus.io.scrape":"false"}
```

**Comment by alexellis** [comment]
The behaviour observed is inconsistent but not technically incorrect so I'll merge and release for the user to test.

**Comment by alexellis** [comment]
@LucasRoesler 

https://github.com/openfaas-incubator/openfaas-operator/pull/48#issuecomment-420725292

We have #50 which is caused by `faas-cli store deploy figlet` (the happy path) - can you confirm you tested this scenario too?

Alex

File: pkg/controller/deployment.go [hunk 1]
```diff
@@ -261,6 +274,19 @@ func envVarsNotEqual(a, b []corev1.EnvVar) bool {
 	return false
 }
 
+func strMapsPtrNotEqual(a, b *map[string]string) bool {
```
Comment (alexellis) [inline pkg/controller/deployment.go]: cc @LucasRoesler @stefanprodan we need tests for this.

`strMapsPtrNotEqual`

and for `strMapsNotEqual` if there are none yet.

**Pull request merged**